### PR TITLE
riscv: add basic SBI implementation

### PR DESCRIPTION
**Pull request by lukileczo**
## Description

SBI handling single & multicore systems and functionality only necessary for Phoenix-RTOS.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [x] Already covered by automatic testing: tested on CI, [works](https://github.com/phoenix-rtos/phoenix-rtos-project/actions/runs/9873029750/job/27264417158)
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work: rebased on top of https://github.com/phoenix-rtos/plo/pull/332
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: riscv-sbi/core/fdt/fdt-internal.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#define FDT_EOK        0
+#define FDT_EBADOFFS   -1
+#define FDT_EBADSTRUCT -2
+#define FDT_EEND       -3
+#define FDT_ENODEEND   -4
+#define FDT_ENOTFOUND  -5
+
+
+#define fdt32_to_cpu(x) (__builtin_bswap32(x))
+#define fdt64_to_cpu(x) (__builtin_bswap64(x))
```
Comment (agkaminski) [inline riscv-sbi/core/fdt/fdt-internal.h:40]: Ok for now, but perhaps endianess of the cpu could be defined in `platform` module?

File: riscv-sbi/devices/uart-16550.c [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+{
+	*(uart16550_common.base + reg) = val;
+	RISCV_FENCE(w, o);
+}
+
+
+static u8 readReg(u8 reg)
+{
+	u8 val = *(uart16550_common.base + reg);
+	RISCV_FENCE(i, r);
```
Comment (agkaminski) [inline riscv-sbi/devices/uart-16550.c:58]: Shouldn't this fence be before reading a value?

File: riscv-sbi/devices/uart-grlib.c [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+/* UART control bits */
+#define RX_EN        (1 << 0)
+#define TX_EN        (1 << 1)
+#define TX_FIFO_FULL (1 << 9)
+
+/* UART status bits */
+#define DATA_READY (1 << 0)
+
+enum {
+	uart_data,   /* Data register           : 0x00 */
```
Comment (agkaminski) [inline riscv-sbi/devices/uart-grlib.c:32]:
```suggestion
	uart_data = 0,   /* Data register           : 0x00 */
```

File: riscv-sbi/include/devices/clint.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+void clint_setTimecmp(u64 time);
+
+
+u64 clint_getTime(void);
+
+
+void clint_init(void);
+
+
+#endif
```
Comment (agkaminski) [inline riscv-sbi/include/devices/clint.h:46]:
```suggestion
#endif

```

File: riscv-sbi/include/devices/console.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+void console_print(const char *s);
+
+
+int console_getc(void);
+
+
+void console_init(void);
+
+
+#endif
```
Comment (agkaminski) [inline riscv-sbi/include/devices/console.h:50]:
```suggestion
#endif

```

File: riscv-sbi/devices/uart-16550.c (hunk 2 above)
Comment (lukileczo) [reply riscv-sbi/devices/uart-16550.c:58]: Both Linux and OpenSBI define MMIO reads with this fence order (and without barrier before read). The read from MMIO will happen before any following reads from main memory which is what we want I guess.

File: riscv-sbi/core/fdt/fdt-internal.h (hunk 1 above)
Comment (lukileczo) [reply riscv-sbi/core/fdt/fdt-internal.h:40]: May as well use the predefined `__BYTE_ORDER__` macro, will do so.

File: riscv-sbi/devices/uart-16550.c (hunk 2 above)
Comment (agkaminski) [reply riscv-sbi/devices/uart-16550.c:58]: Ok, for some reason I was confused and thought that this particular example was in different order than other instances, but it's not.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**